### ISSUE TITLE
Compilation failed on Centos 7 system when building with mainline DPDK

### ISSUE THREAD

**Issue by karlatec** [spdk/spdk#1891]
## Current Behavior

SPDK fails to build on a Centos 7 system when  building with custom DPDK (`--with-dpdk`) due to DPDK compilation issue.
Example failed build: https://ci.spdk.io/public_build/autotest-spdk-master-vs-dpdk-main_1501.html

First failed occurence: 
https://ci.spdk.io/results/autotest-spdk-master-vs-dpdk-main/builds/1498/archive/centos7-nightly-vg-autotest/index.html
DPDK `340d22c event/octeontx2: fix device reconfigure for single slot`

## Steps to Reproduce

1. Create autorun-spdk.conf file:
```
SPDK_RUN_FUNCTIONAL_TEST=1
SPDK_TEST_BLOCKDEV=1
SPDK_RUN_ASAN=1
RUN_NIGHTLY=0
SPDK_TEST_NATIVE_DPDK=main
SPDK_RUN_EXTERNAL_DPDK=/spdk-temp/dpdk/build
```
2. Run `spdk/autorun.sh autorun-spdk.conf`
3. Build fails at:
```
00:03:11.791  FAILED: lib/librte_eventdev.a.p/librte_eventdev_rte_event_eth_rx_adapter.c.o 
00:03:11.791  cc -Ilib/librte_eventdev.a.p -Ilib -I../lib -Ilib/librte_eventdev -I../lib/librte_eventdev -I. -I.. -Iconfig -I../config -Ilib/librte_eal/include -I../lib/librte_eal/include -Ilib/librte_eal/linux/include -I../lib/librte_eal/linux/include -Ilib/librte_eal/x86/include -I../lib/librte_eal/x86/include -Ilib/librte_eal/common -I../lib/librte_eal/common -Ilib/librte_eal -I../lib/librte_eal -Ilib/librte_kvargs -I../lib/librte_kvargs -Ilib/librte_metrics -I../lib/librte_metrics -Ilib/librte_telemetry -I../lib/librte_telemetry -Ilib/librte_ring -I../lib/librte_ring -Ilib/librte_ethdev -I../lib/librte_ethdev -Ilib/librte_net -I../lib/librte_net -Ilib/librte_mbuf -I../lib/librte_mbuf -Ilib/librte_mempool -I../lib/librte_mempool -Ilib/librte_meter -I../lib/librte_meter -Ilib/librte_hash -I../lib/librte_hash -Ilib/librte_rcu -I../lib/librte_rcu -Ilib/librte_timer -I../lib/librte_timer -Ilib/librte_cryptodev -I../lib/librte_cryptodev -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -include rte_config.h -Wextra -Wcast-qual -Wdeprecated -Wformat -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wold-style-definition -Wpointer-arith -Wsign-compare -Wstrict-prototypes -Wundef -Wwrite-strings -Wno-missing-field-initializers -D_GNU_SOURCE -fPIC -g -Werror -fcommon -Wno-stringop-overflow -fPIC -march=native -DALLOW_EXPERIMENTAL_API -DALLOW_INTERNAL_API -DLINUX -MD -MQ lib/librte_eventdev.a.p/librte_eventdev_rte_event_eth_rx_adapter.c.o -MF lib/librte_eventdev.a.p/librte_eventdev_rte_event_eth_rx_adapter.c.o.d -o lib/librte_eventdev.a.p/librte_eventdev_rte_event_eth_rx_adapter.c.o -c ../lib/librte_eventdev/rte_event_eth_rx_adapter.c
00:03:11.791  ../lib/librte_eventdev/rte_event_eth_rx_adapter.c: In function â€˜rxa_set_vector_dataâ€™:
00:03:11.791  ../lib/librte_eventdev/rte_event_eth_rx_adapter.c:1810:57: error: signed and unsigned type in conditional expression [-Werror=sign-compare]
00:03:11.791     flow_id == 0 ? (qid & 0xFFF) | (port_id & 0xFF) << 12 : flow_id;
00:03:11.791                                                           ^
00:03:11.791  ../lib/librte_eventdev/rte_event_eth_rx_adapter.c: At top level:
00:03:11.791  cc1: error: unrecognized command line option "-Wno-stringop-overflow" [-Werror]
00:03:11.791  cc1: all warnings being treated as errors
00:03:11.852  [248/693] Generating hash.sym_chk with a custom command (wrapped by meson to capture output)
00:03:11.915  [249/693] Generating symbol file lib/librte_timer.so.21.2.p/librte_timer.so.21.2.symbols
00:03:12.007  [250/693] Generating symbol file lib/librte_rcu.so.21.2.p/librte_rcu.so.21.2.symbols
00:03:12.024  [251/693] Generating symbol file lib/librte_mempool.so.21.2.p/librte_mempool.so.21.2.symbols
00:03:12.039  [252/693] Generating symbol file lib/librte_acl.so.21.2.p/librte_acl.so.21.2.symbols
00:03:12.090  [253/693] Generating efd.sym_chk with a custom command (wrapped by meson to capture output)
00:03:12.203  [254/693] Compiling C object lib/librte_eventdev.a.p/librte_eventdev_rte_event_timer_adapter.c.o
00:03:12.257  [255/693] Generating ethdev.sym_chk with a custom command (wrapped by meson to capture output)
00:03:12.925  [256/693] Compiling C object lib/librte_eventdev.a.p/librte_eventdev_rte_event_crypto_adapter.c.o
00:03:12.925  ninja: build stopped: subcommand failed.
```

## Context (Environment including OS version, SPDK version, etc.)

centos7-cloud-210126160 3.10.0-1160.11.1.el7.x86_64
Meson 0.56.2
ninja 1.10.0
gcc 4.8.5 20150623 (Red Hat 4.8.5-44)

**Comment by jimharris** [comment]
DPDK says they only support gcc >= 4.9:

http://doc.dpdk.org/guides/linux_gsg/sys_reqs.html

**Comment by mikeBashStuff** [comment]
I am aware that centos7 will reach its EOL in 2024, but this distro is really lost in time...

**Comment by karlatec** [comment]
Oh... so I guess we now have to decide how to proceed from here:
* Provision our Centos7 VMs with custom-build, recent gcc (I highly suggest we don't do that :) )
* Remove Centos7 from vs-dpdk tests
* Stop testing with Centos7 completely.

I think we should just remove it from vs-dpdk tests now, and stop testing with it completely after 21.04 release. At some point when we will want to update dpdk submodule, and per-patch tests will fail too then.

**Comment by jimharris** [comment]
This -Wno-stringop-overflow option is coming from SPDK (dpdkbuild/Makefile) not DPDK.  See commit 970c6d099e which was merged shortly before 20.07 release.

@karlatec - any thoughts why we would only be seeing this issue now?

**Comment by jimharris** [comment]
Also note that dpdkbuild/Makefile is trying to only enable this option on gcc > 10.x - but that check seems to be failing here.  Maybe the gcc output was different enough in gcc 4.8 that it confuses our check.

**Comment by tomzawadzki** [comment]
[Bug scrub] Latest builds on mainline DPDK with Centos7 look to be passing, @karlatec will confirm that the issue reported here no longer occurs.

**Comment by karlatec** [comment]
I take that back. It's failing again: https://ci.spdk.io/public_build/autotest-spdk-master-vs-dpdk-main_1532.html

Here's the git log diff between last CentOS7 [passing build](https://ci.spdk.io/public_build/autotest-spdk-master-vs-dpdk-main_1526.html) and [first failing build](https://ci.spdk.io/public_build/autotest-spdk-master-vs-dpdk-main_1527.html):
```
examples/l3fwd: fix LPM IPv6 subnets
examples/l3fwd: move routes to common header
examples/l3fwd: add FIB infrastructure
examples/l3fwd: implement FIB lookup method
pipeline: modularize SWX instruction optimizer <--- this seems to match the fail
pipeline: auto-detect endianness of action arguments
```

@jimharris maybe we should consider removing CentOS7 from vs-dpdk tests? If they officially support only 4.9 and higher then I think fails like these will happen frequently.

**Comment by jimharris** [comment]
It looks like the uninitialized warnings were fixed in upstream:

http://git.dpdk.org/dpdk/commit/lib?id=fe4b8c7bcdba8c81b77e5c28b0b55c5a6653a3c6

So I guess DPDK is still trying to support gcc 4.8.5.

But I'm still seeing this in the log:
00:03:11.791  cc1: error: unrecognized command line option "-Wno-stringop-overflow" [-Werror]

See my other note above, this option is passed by SPDK to the DPDK build, but it is supposed to only pass it on gcc>10.x.  Can you run this on the CentOS 7 system and paste the output here?

echo __GNUC__ | gcc -E -x c -

**Comment by jimharris** [comment]
I think for now we should leave the CentOS 7 tests, maybe we can modify dpdkbuild/Makefile to detect gcc 4.8 and pass -Wno-error so we can reduce the number of failures like this.

**Comment by karlatec** [comment]
```
sys_sgci@poweredge2:~/centos7-libvirt$ echo __GNUC__ | gcc -E -x c -
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 31 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 32 "<command-line>" 2
# 1 "<stdin>"
7
```

**Comment by jimharris** [comment]
That's looks right.  I don't understand why the dpdkbuild/Makefile checks aren't working then.  You may need to add some echo statements to the 'all' target, i.e. echo $(GCC_MAJOR).  Maybe also add:

```
# add this after GCC_MAJOR is assigned
GCC_TEST=$(shell test $(GCC_MAJOR) -ge 10 && echo 1)

# add this to the 'all' target
  echo $(GCC_TEST)
```

**Comment by karlatec** [comment]
I know why the version does not get detected. Because we build external DPDK via autobuild.sh / build_native_dpdk function. dpdkbuild/Makefile is not used yet.
This should fix it: https://review.spdk.io/gerrit/c/spdk/spdk/+/7536

Also this https://review.spdk.io/gerrit/c/spdk/spdk/+/7537/1 for -Wno-stringop-overflow

**Comment by jimharris** [comment]
Oh of course, that makes so much more sense now.  Both of those patches look good to me.

**Comment by karlatec** [comment]
https://ci.spdk.io/public_build/autotest-spdk-master-vs-dpdk-main_1546.html
Results should be available in about 1-1.5 hour. The test is still running. Posting the comment now, because I'm wrapping up for today :)

**Comment by jimharris** [comment]
Looks good.  Closing.
